File: lab2/5.py
#define max(a,b) ((a >= b) ? a : b)
#include <stdlib.h>
#include <stdio.h>
void printArr(int a[],  int n){
    for (int i = 0; i < n; i++){
        printf("%d ", a[i]);
    }
    printf("\n");
}
 
void qsort2(int a[], long long l, long long r, long long k){
    if (l >= r) return;
    long long mid = a[(l + r) / 2];
    long long i = l, j = r, temp;
    while (i <= j){
        while (a[i] < mid)
            i++;
        while (a[j] > mid)
            j--;
        if (i <= j){
            temp = a[i];
            a[i] = a[j];
            a[j] = temp;
            i++;
            j--;
        }
    }
    if (k <= j) return qsort2(a, l, j, k);
    else return qsort2(a, i, r, k);
}
 
int main() {
    long long A, B, C, n, k;
    FILE * fin = fopen("kth.in", "r");
    FILE * fout = fopen("kth.out", "w");
 
    fscanf(fin, "%lld", &n);
    fscanf(fin, "%lld", &k);
    fscanf(fin, "%lld", &A);
    fscanf(fin, "%lld", &B);
    fscanf(fin, "%lld", &C);
    int *a = malloc(n*sizeof(int));
    fscanf(fin, "%d", &a[0]);
    fscanf(fin, "%d", &a[1]);
 
    for (long long i = 2; i < n; i++)
        a[i] = A*a[i-2] + B * a[i-1] + C;
 
    qsort2(a, 0, n-1, k-1);
 
    fprintf(fout, "%d", a[k-1]);
}

File: lab4/5.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

typedef struct node{
    int value;
    struct node* next;
}node;

void push(node** head, int data){
    node* tmp = malloc(sizeof(node));
    tmp->value = data;
    tmp->next = *head;
    (*head) = tmp;
}

int pop(node** head){
    if (*head == NULL){
        printf("Error: list is empty");
        exit(-1);
    }
    int res = (**head).value;
    *head = (*head)->next;
    return res;
}

node* getSecondLast(node*head){
    if (head == NULL)
        exit(-2);

    if (head->next == NULL)
        return NULL;

    while (head->next->next)
        head = head->next;

    return head;
}

int popBack(node** head) {
    int res = 0;
    node *sLast = NULL;
    if (!head || !(*head))
        exit(-1);
    sLast = getSecondLast(*head);

    if (sLast == NULL){
        res = (*head)->value;
        free(*head);
        *head = NULL;
    }else{
        res = sLast->next->value;
        free(sLast->next);
        sLast->next = NULL;
    }
    return res;
}

int main(){
    node *head = NULL;

    FILE * fin = fopen("postfix.in", "r");
    FILE * fout = fopen("postfix.out", "w");
    int temp;
    char str[10000];
    while(fscanf(fin, "%s", str) != EOF) {
        if (strcmp(str, "+") == 0) {
            temp = head->value + head->next->value;
            pop(&head);
            pop(&head);
            push(&head, temp);
        }
        else if (strcmp(str, "*") == 0) {
            temp = head->value * head->next->value;
            pop(&head);
            pop(&head);
            push(&head, temp);
        }
        else if (strcmp(str, "-") == 0) {
            temp = head->next->value - head->value;
            pop(&head);
            pop(&head);
            push(&head, temp);
        }
        else
            push(&head, atoi(str));
    }
    fprintf(fout, "%d", pop(&head));
}